Webhooks guide security (#585)
liveblocks/liveblocks: docs/pages/guides/webhooks.mdx

available for subscription, along with their schema.

Events available for use include:

- UserEntered/UserLeft
- StorageUpdated

> - RoomCreated/RoomDeleted

Reviewer: @FlowFlorent do we really support those events?

- MaxConnectionsReached

#### UserEntered

```ts
// Schema
type UserEntered = {
	type: "userEntered"
	data: {
		appId: string
		roomId: string
	  connectionId: number
		userId: string | null
		userInfo: Record<string, any> | null
		enteredAt: string
		numActiveUsers: number
	}
}

// Example
const userEnteredMessage = {
	type: "userEntered",
	data: {
		appId: "aasdbe234bsdf"
		roomId: "my-room-1"
		connectionId: 4
		userId: '1234sdgs'
		userInfo: null
		enteredAt: '2021-10-06T01:45:56.558Z'
		numActiveUsers: 8
	}
}
```

> #### UserLeft

Reviewer: Could we have a short description of what it is in human language?

Author: good catch, added

available for subscription, along with their schema.

Events available for use include:

- UserEntered/UserLeft
- StorageUpdated

More events will come later such as:

- RoomCreated/RoomDeleted
- MaxConnectionsReached

#### UserEntered

> A user enters a room when they connect to a room, which is when this event is triggered. The number of active users is the number of users in the room after the
> user has entered.

Reviewer: ```suggestion
A user enters a room when they connect to a room, which is when this event is triggered. The `numActiveUsers` field represents the number of users in the room 
after the user has entered.
```

```ts
// Schema
type UserEntered = {
  type: "userEntered";
  data: {
    appId: string;
    roomId: string;
    connectionId: number;
    userId: string | null;
    userInfo: Record<string, any> | null;
    enteredAt: string;
    numActiveUsers: number;
  };
};

// Example
const userEnteredMessage = {
  type: "userEntered",
  data: {
    appId: "my-app-id",
    roomId: "my-room-id",
    connectionId: 4,
    userId: "a-user-id",
    userInfo: null,
    enteredAt: "2021-10-06T01:45:56.558Z",
    numActiveUsers: 8,
  },
};
```

#### UserLeft

> A user leaves a room when they disconnect from a room, which is when this event is triggered. The number of active users is the number of users in the room 
> after the user has left.

Reviewer: ```suggestion
A user leaves a room when they disconnect from a room, which is when this event is triggered. The `numActiveUsers` field represents the number of users in the room 
after the user has left.
```

available for subscription, along with their schema.

Events available for use include:

- UserEntered/UserLeft
- StorageUpdated

More events will come later such as:

- RoomCreated/RoomDeleted
- MaxConnectionsReached

#### UserEntered

> A user enters a room when they connect to a room. This event is triggered when a
> user enters a room. The number of active users is the number of users after the
> user has entered.

Reviewer: ```suggestion
A user enters a room when they connect to a room, which is when this event is triggered. The number of active users is the number of users in the room after the
user has entered.
```

```ts
// Schema
type UserEntered = {
  type: "userEntered";
  data: {
    appId: string;
    roomId: string;
    connectionId: number;
    userId: string | null;
    userInfo: Record<string, any> | null;
    enteredAt: string;
    numActiveUsers: number;
  };
};

// Example
const userEnteredMessage = {
  type: "userEntered",
  data: {
    appId: "my-app-id",
    roomId: "my-room-id",
    connectionId: 4,
    userId: "a-user-id",
    userInfo: null,
    enteredAt: "2021-10-06T01:45:56.558Z",
    numActiveUsers: 8,
  },
};
```

#### UserLeft

> A user leaves a room when they disconnect from a room. This event is triggered
> when a user leaves a room. The number of active users is the number of users
> after the user has left.

Reviewer: ```suggestion
A user leaves a room when they disconnect from a room, which is when this event is triggered. The number of active users is the number of users in the room 
after the user has left.
```

```ts
// Schema
type UserLeft = {
  type: "userLeft";
  data: {
    appId: string;
    roomId: string;
    connectionId: number;
    userId: string | null;
    userInfo: Record<string, any> | null;
    leftAt: string;
    numActiveUsers: number;
  };
};

// Example
const userLeftMessage = {
  type: "userLeft",
  data: {
    appId: "my-app-id",
    roomId: "my-room-id",
    connectionId: 4,
    userId: "a-user-id",
    userInfo: {
      name: "John Doe",
    },
    leftAt: "2021-10-06T01:45:56.558Z",
    numActiveUsers: 7,
  },
};
```

#### StorageUpdated

> Storage is updated when a user writes to storage. This event is throttled and as
> such may not be triggered for every write. The `updatedAt` field is the time of
> the last write.

Reviewer: ```suggestion
Storage is updated when a user writes to storage. This event is throttled and, as
such, may not be triggered for every write. The `updatedAt` field is the time of
the last write.
```

service incurs considerable downtime (over 10 hours), you can replay individual
messages from the Endpoints portion of the dashboard by clicking the kebab menu
on an individual message, or you can opt to bulk replay events by clicking the
main menu and selecting "Replay Failed Messages."

<Figure>
  <video autoPlay loop muted playsInline>
    <source src="/images/docs/webhooks/replay-events.mp4" type="video/mp4" />
  </video>
</Figure>

Each message is attempted based on the following schedule, where each period is
started following the failure of the preceding attempt:

- Immediately
- 5 seconds
- 5 minutes
- 30 minutes
- 2 hours
- 5 hours
- 10 hours
- 10 hours (in addition to the previous)

> If an endpoint is removed or disabled delivery attempts to the endpoint will be

Reviewer: ```suggestion
If an endpoint is removed or disabled, delivery attempts to the endpoint will be
```

When you verify your webhooks, you protect against man-in-the-middle, CSRF, and
replay attacks. Because of this, the importance of verification cannot be
understated and should be considered paramount in your integration.

Here's how to validate a webhook message:

#### 1. Construct the signed content

The content to sign is composed by concatenating the id, timestamp and payload
of the request, separated by the full-stop character (`.`). In code, it will
look something like:

```ts
const crypto = require("crypto");

// webhookId comes from the `webhook-id` header
// webhookTimestamp comes from the `webhook-timestamp` header
// body is the request body
signedContent = `${webhookId}.${webhookTimestamp}.${body}`;
```

#### 2. Generate the signature

```ts
// Your endpoint's signing secret
const secret = "whsec_5WbX5kEWLlFzsG3jH64I8loOqUB6e8FH";

// Need to base64 decode the secret
const secretBytes = new Buffer(secret.split("_")[1], "base64");
// This is the signature you will compare against the signature header
const signature = crypto
  .createHmac("sha256", secretBytes)
  .update(signedContent)
  .digest("base64");
```

> #### 2. Validate the signature to the signature header

Author: ```suggestion
#### 2. Validate the signature
```
